### PR TITLE
Protobuf should be serialized to direct memory instead of heap memory

### PR DESCRIPTION
**Pull request by karanmehta93**
Descriptions of the changes in this PR:

### Motivation

Protobuf serialization is the last step of the netty pipeline. Currently, we allocate a heap buffer while serializing
and pass it down to netty library. In `AbstractChannel#filterOutboundMessage()`, netty copies that data to a direct buffer if
it is currently in heap (otherwise skips it and uses it directly).
This change helps in reducing unncessary CPU cycles for buffer copies and also helps alleviate pressure off the GC, since there is
less churn of memory. In some stress tests, I see a reduction upto 33% in number of GC's with the same avg running duration in BK Client.
The workload does writes of different sizes, ranging from 1 KB to 512 KB.
Bookies aren't usually CPU bound. This change improves READ_ENTRY code paths by a small factor.

### Changes

Used direct memory for buffer allocation versus heap memory

### CODE REVIEW

**Comment by karanmehta93** [comment]
@eolivelli @sijie Some improvements to our netty stack. It would be helpful if you can run a perf test with these changes in your environments before pushing it in.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieProtoEncoding.java [hunk 1]
```diff
@@ -356,10 +356,10 @@ public Object encode(Object msg, ByteBufAllocator allocator) throws Exception {
 
     private static ByteBuf serializeProtobuf(MessageLite msg, ByteBufAllocator allocator) {
         int size = msg.getSerializedSize();
-        ByteBuf buf = allocator.heapBuffer(size, size);
+        ByteBuf buf = allocator.directBuffer(size, size);
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieProtoEncoding.java:359]: shall we add a comment that explains why this is better ?

**Comment by eolivelli** [comment]
@merlimat PTAL as well

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieProtoEncoding.java (hunk 1 above)
Comment (karanmehta93) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieProtoEncoding.java:359]: I will do it. For reviewers: it is the same as one in PR description

**Comment by Ghatage** [comment]
rerun failure checks

**Comment by eolivelli** [comment]
All tests passed.
I wonder if someone has cycles to perform benchmarks with this patch

cc @zymap @jiazhai @ravisharda

**Review by eolivelli — APPROVED** [review]
LGTM

**Review by jiazhai — APPROVED** [review]

**Pull request merged**